plugin-server/src/main/ingestion-queues/session-recording/services/session-manager-v3.ts

```diff
@@ -0,0 +1,472 @@
+    help: 'The size of sessions in numbers of lines',
+    buckets: BUCKETS_LINES_WRITTEN,
+})
+
+const writeStreamBlocked = new Counter({
+    name: metricPrefix + 'recording_blob_ingestion_write_stream_blocked',
+    help: 'Number of times we get blocked by the stream backpressure',
+})
+
+export type SessionManagerBufferContext = {
+    sizeEstimate: number
+    count: number
+    eventsRange: {
+        firstTimestamp: number
+        lastTimestamp: number
+    } | null
+    createdAt: number
+}
+
+export type SessionBuffer = {
+    context: SessionManagerBufferContext
+    fileStream: Transform
+}
+
+// Context that is updated and persisted to disk so must be serializable
+export type SessionManagerContext = {
+    dir: string
+    sessionId: string
+    teamId: number
+    partition: number
+}
+
+export class SessionManagerV3 {
+    buffer?: SessionBuffer
+    flushPromise?: Promise<void>
+    destroying = false
+    inProgressUpload: Upload | null = null
+    flushJitterMultiplier: number
+
+    constructor(
+        public readonly serverConfig: PluginsServerConfig,
+        public readonly s3Client: ObjectStorage['s3'],
+        public readonly context: SessionManagerContext
+    ) {
+        // We add a jitter multiplier to the buffer age so that we don't have all sessions flush at the same time
+        this.flushJitterMultiplier = 1 - Math.random() * serverConfig.SESSION_RECORDING_BUFFER_AGE_JITTER
+    }
+
+    private file(name: string): string {
+        return path.join(this.context.dir, name)
+    }
+
+    public static async create(
+        serverConfig: PluginsServerConfig,
+        s3Client: ObjectStorage['s3'],
+        context: SessionManagerContext
+    ): Promise<SessionManagerV3> {
+        const manager = new SessionManagerV3(serverConfig, s3Client, context)
+        await mkdir(context.dir, { recursive: true })
+
+        try {
+            const bufferMetadata: SessionManagerBufferContext = JSON.parse(
+                await readFile(path.join(context.dir, 'metadata.json'), 'utf-8')
+            )
+            manager.buffer = {
+                context: bufferMetadata,
+                fileStream: manager.createFileStreamFor(path.join(context.dir, BUFFER_FILE_NAME)),
+            }
+        } catch (error) {
+            // Indicates no buffer metadata file or it's corrupted
+            status.error('🧨', '[session-manager] failed to read buffer metadata', {
+                ...context,
+                error,
+            })
+        }
+
+        status.info('📦', '[session-manager] started new manager', {
+            ...manager.context,
+            ...(manager.buffer?.context ?? {}),
+        })
+        return manager
+    }
+
+    private async save(): Promise<void> {
+        if (this.buffer) {
+            await writeFile(this.file('metadata.json'), JSON.stringify(this.buffer?.context), 'utf-8')
+        } else {
+            await unlink(this.file('metadata.json'))
```

Comment:
deleting a file in save is confusing (to my tired brain)... what's going on here?

Reply:
This is syncing the state to disk in the metadata file. So its called whenever something changes with the buffered data.